### DISCUSSION TITLE
Non local stress update

### DISCUSSION THREAD

**Discussion by albiedahal** [idaholab/moose#19003] [Q&A General] [answered]
Hi all,

I am trying to implement a nonlocal return mapping algorithm for inelastic stress update as per [this paper](https://drive.google.com/file/d/1-CrobvlOFKmvClbIJzerq5R4MAXYdrYB/view?usp=sharing). The figure attached outlines the stress update algorithm for the nonlocal approach. 

![image](https://user-images.githubusercontent.com/55407326/135933371-61eb6308-0955-4b00-9f30-1d9b5e35e174.png)

The values of the plastic multiplier for all of the quadrature points are coupled and calculated simultaneously (Ref Equations 33, 34, and 41). As far as I know, traditional (local) stress update materials in MOOSE compute plastic strain at a quadrature point independent of the stress state of the other quadrature points, which does not work for this approach. Except for the return map, the solution method is no different from the standard finite element method, so I need to update the stresses at the quadrature points globally and store those values for the kernels to consume.  Any pointers or suggestions on how and where to start would be highly appreciated. 

Kind regards,

Albert

**Comment by jiangwen84** [comment, marked as answer]
which paper?

**Reply by albiedahal** [reply]
Sorry for that. I have linked the paper now.

**Reply by hugary1995** [reply]
So in MOOSE we loop over elements to compute material properties, e.g. stress and strain.

The method called at each element is `computeProperties`. Within this method, we loop over quadrature points, and at each quadrature point, we call `computeQpProperties`. So you do have access to this sort of non-local information within the `computeProperties` method. We do this routinely for stuff like volumetric locking correction. You can take a look at `ComputeSmallStrain::computeProperties()` for example.

The stress update classes derive from `StressUpdateBase`, which has the method `setQp` that allows you to set the current quadrature point. This could also be useful when implementing this non-local method.

**Reply by jiangwen84** [reply]
@albiedahal    Thanks for the paper. Unfortunately, we do not have the access to that paper :-( 

Does `all integration points` in this algorithm refer to the quadrature points within a single element or all elements? If it is for a single element, @hugary1995 's suggestion is what you should follow. If it evolves multiple elements, we need to read the paper and give you other suggestions.

**Reply by albiedahal** [reply]
@hugary1995 Thank you for the reply. I need to loop over the quadrature points of all the elements in the domain, not just at a particular element. I don't think `computeProperties` allows me to do that?

@jiangwen84 I have updated the link to a google drive link which should be accessible now. The `all integration points` in the algorithm refers to all the quadrature points in the domain over all elements.

**Reply by GiudGiud** [reply]
@albiedahal app developers usually write an implementation for computeQpProperties. But if you want to compute the matprop at all qps together, you can re-implement computeProperties. It s in Material.C it s not very complicated
EDIT this is a little hacky, hence the dislike

**Reply by albiedahal** [reply]
@GiudGiud Thank you, I'll definitely look into that. I am having a hard time figuring the qp indexes and the prop ids in the `computeProperties` as I do not have much programming background/experience.

**Comment by dschwen** [comment]
Can you explain what the "nonlocal internal variables" are?

**Reply by albiedahal** [reply]
@dschwen the non local variables in this algorithm is the non local plastic strain increment (or the plastic multiplier)  `$\tilde{\kappa}$`. The non local plastic strain increment for ith quadrature point in the domain includes weighted contributions of the local plastic strains `$\kappa$` of all quadrature points in the domain (Eq. 33). The `a_ij` in is the non-local weight and `w_j` is the quadrature rule weight. The yield function (Eq. 34) in this case is not satisfied explicitly at each quadrature points locally, but more in a distributed sense by specifying a tolerance `$\delta_f$`.
